logging/journald_core.go

```diff
@@ -124,3 +132,61 @@ func encodeJournaldFieldKey(key string) string {
 
 	return key
 }
+
+// visibleFieldsMsg creates a string to be appended to the log message including fields to be explicitly printed.
+//
+// When logging against journald, the zapcore.Fields are used as journald fields, resulting in not being shown in the
+// default journalctl output (short). While this is documented in our docs, missing error messages are usually confusing
+// for end users.
+//
+// This method takes an allow list (set, map of keys to empty struct) of key to be displayed - there is the global
+// variable journaldVisibleFields; parameter for testing - and a slice of zapcore.Fields, creating an output string of
+// the allowed fields prefixed by a whitespace separator. If there are no fields to be logged, the returned string is
+// empty. So the function output can be appended to the output message without further checks.
+func visibleFieldsMsg(visibleFieldKeys map[string]struct{}, fields []zapcore.Field) string {
+	if visibleFieldKeys == nil || fields == nil {
+		return ""
+	}
+
+	enc := zapcore.NewMapObjectEncoder()
+
+	for _, field := range fields {
+		if _, shouldLog := visibleFieldKeys[field.Key]; shouldLog {
+			field.AddTo(enc)
+		}
+	}
+
+	// The internal zapcore.encodeError function[^0] can result in multiple fields. For example, an error type
+	// implementing fmt.Formatter results in another "errorVerbose" field, containing the stack trace if the error was
+	// created by github.com/pkg/errors including a stack[^1].
+	//
+	// Therefore, the keys are checked again in the following loop before being inserted in the final array, having a
+	// capacity of twice the fields length to minimize the chance of a resize.
+	//
+	// [^0]: https://github.com/uber-go/zap/blob/v1.27.0/zapcore/error.go#L47
+	// [^1]: https://pkg.go.dev/github.com/pkg/errors@v0.9.1#WithStack
+	visibleFields := make([]string, 0, 2*len(enc.Fields))
+	for k, v := range utils.IterateOrderedMap(enc.Fields) {
+		if _, shouldLog := visibleFieldKeys[k]; !shouldLog {
+			continue
+		}
+
+		var fieldVerb, fieldEnclosure string
+		switch v.(type) {
+		case rune, string, []byte, error:
+			fieldVerb = "%q"
+			// %q already enclosures the value
+		default:
+			fieldVerb = "%v"
+			fieldEnclosure = `"`
+		}
+
+		visibleFields = append(visibleFields, fmt.Sprintf("%s=%s"+fieldVerb+"%s", k, fieldEnclosure, v, fieldEnclosure))
```

Comment:
Don't you end up with exactly two different format strings here in the end? Wouldn't it be much simpler to just specify both of them explicitly?

Reply:
I have merged them into one format string. Please take another look.

Reply:
I was referring to the whole format string, just having the whole string there is just easier to read, even if you repeat the `%s=`.

I'd even go so far and move the `fmt.Sprintf()` into the different cases as well (that should make it easier for linters):

```go
		var encodedField string
		switch v.(type) {
		case rune, string, []byte, error:
			encodedField = fmt.Sprintf("%s=%q", k, v)
		default:
			encodedField = fmt.Sprintf(`%s="%v"`, k, v)
		}

		visibleFields = append(visibleFields, encodedField)
```

Reply:
That even makes it easier. I have pushed a new version.